Fix build using system boost/ system yaml-cpp

dvzrv opened the pull request:
Purpose and Motivation
----------------------

Some components (`editors/sc-ide`, `testsuite/server`, `yaml-cpp`, `oscpack`) unconditionally use the `external_libraries/boost` in `target_include_dirs` leading to problems when trying to build against system versions of boost and yaml-cpp (#4096).
This pull request unifies the naming scheme for referencing boost libraries within `CMakelists.txt` to `boost_c_lib` and sets Boost's include directory to `boost_include_dirs` depending on `SYSTEM_BOOST`.

Types of changes
----------------

- Bug fix (non-breaking change which fixes an issue)

Checklist
---------

- [x] All previous tests are passing
- ~~[ ] Tests were updated or created to address changes in this PR, and tests are passing~~
- [x] Updated documentation, if necessary
- [x] This PR is ready for review

The tests sadly don't really apply, as some don't work headlessly or generally (#3555).

Remaining Work
--------------

Travis integration to test with/without system boost/ yaml will be done in a separate pull request

brianlheim commented:
Thanks for the PR @dvzrv . I can't review it in full tonight but just want to leave some initial comments in case you have some time to update it:

- please don't do drive-by formatting (for reasons I explained in the sc3-plugins PR)
- use 4 space indents when it doesn't break with surrounding style. this will soon be normalized across all project files and if you add 2-space code now, it's only going to be obliterated in git blame
- i think we should use SYSTEM_BOOST rather than Boost_FOUND to switch on conditions. The latter is less under our control and doesn't express the actual condition we're interested in
- for the oscpack change - you could leave that where it is, combine both calls to `target_include_directories` and use a [generator expression](https://cmake.org/cmake/help/v3.5/manual/cmake-generator-expressions.7.html) to add boost under the right condition. that would be much cleaner

dvzrv commented:
I undid my previous unnecessary commits.
I have changed all conditionals, previously relying on `Boost_FOUND` to rely on `SYSTEM_BOOST` instead. This now gives the possibility of a cleaner check and a sane fallback to the vendored boost, if no system boost was found.

brianlheim commented on external_libraries/CMakeLists.txt:
```diff
@@ -104,7 +105,11 @@ if(NOT YAMLCPP_FOUND)
+    target_include_directories(yaml PUBLIC ${Boost_INCLUDE_DIRS})
+  else()
+    target_include_directories(yaml PUBLIC ${CMAKE_SOURCE_DIR}/external_libraries/yaml-cpp/include boost)
+  endif()
```
in both cases, you need the `external_libraries` include, otherwise building fails if you:
- have yaml-cpp in a reachable directory in an older version
- use SYSTEM_BOOST
- don't use SYSTEM_YAMLCPP

this happened to me :)

dvzrv replied:
Wouldn't this also mean unconditionally including the local version of yaml-cpp?

brianlheim replied:
the message at the top of this block says "using bundled libyaml-cpp", so i'm assuming that is intended here.

dvzrv commented on editors/sc-ide/CMakeLists.txt:
```diff
@@ -299,11 +298,12 @@ target_link_libraries( SuperCollider
 )
 
-if(Boost_FOUND)
+if(${SYSTEM_BOOST})
```
I'm writing too much bash these days.... -_-

brianlheim replied:
too much bash? no such thing

brianlheim commented:
@dvzrv - tested this on my system today and i run into problems with building supernova because I don't have a non-multithreaded version of `boost_thread` installed (`libboost_thread` as opposed to `libboost_thread-mt`). supernova and its test targets request to link with `boost_thread` without doing the switch on SYSTEM_BOOST used elsewhere: 

https://github.com/supercollider/supercollider/blob/ce76469ed52ad455267803f90cb0da37c0983dd1/testsuite/server/supernova/CMakeLists.txt#L37

i think this points to a larger problem, which is that right now the logic for what it means to link with "boost" is not contained well in these scripts, so everything that wants to link with boost always needs this `if()` block boilerplate. moreover, it leaves ambiguity about which versions of libraries are being used, single- or multi-threaded.

i think we should contain these in separate cmake variables so that calls can look like 

```cmake
target_include_directories(targetname PUBLIC ${real_boost_include})
target_link_libraries(targetname PUBLIC ${real_boost_thread} ${real_boost_filesystem})
```

instead of

```cmake
if(SYSTEM_BOOST)
    target_include_directories(...)
    target_link_libraries(...)
else()
    target_include_directories(...)
    target_link_libraries(...)
endif()
```

i wasn't going to suggest this in this PR because it wasn't strictly necessary, but now that more changes are needed, i don't see the point in maintaining obviously bad boilerplate.

one idea is to set `Boost_INCLUDE_DIRS`, `Boost_THREAD_LIBRARY`, etc when we are building boost targets ourselves: https://github.com/supercollider/supercollider/blob/ce76469ed52ad455267803f90cb0da37c0983dd1/external_libraries/CMakeLists.txt#L5

then everything can just reference those directly. i.e. a spot like this: (https://github.com/supercollider/supercollider/blob/ce76469ed52ad455267803f90cb0da37c0983dd1/server/scsynth/CMakeLists.txt#L179-L183)

can become:

```cmake
target_link_libraries(libscsynth ${Boost_SYSTEM_LIBRARY} ${Boost_FILESYSTEM_LIBRARY} )
```

i don't know if i feel good about using boost-specific variable names for this, but i can't think of a way that this would muck anything up, even if you switched `SYSTEM_BOOST` on/off between successive cmake configurations. but i'm fine being superstitious as well, separate variable names would also be fine ;D they just need to be set clearly, in the same bit of code.

i think if this logic is simplified it will be a lot easier to reason about the build system overall. what do you think?

brianlheim commented:
> it would mean reusing the canonical Boost lib names, as set by FindBoost by changing the add_library() calls within this conditional statement in external_libraries/CMakeLists.txt?

Yes. do you think that's a good idea? maybe it will be misleading. i also have only about 70% confidence that it won't cause trouble to use those names.

> What will be done in the case that something links directly against boost, such as yaml in your review above?

I'm not sure what you mean... do you mean link against boost or use its include directory? the comment i left about is about a call to `target_include_directories` afaict

dvzrv commented:
> Yes. do you think that's a good idea? maybe it will be misleading. i also have only about 70% confidence that it won't cause trouble to use those names.

Well, that doesn't sound too great. ;-)
However, I like the unifying approach and this will make things more readable, but I really don't know what will happen, if implemented (as it needs build tests against several targets - which should also be done by travis imho) and specifically, if this should be done in this pull request as well.

Closing: Is it possible to set new lib names (e.g. `${sc_boost_system}`) for the system libraries using `add_library()`, instead of reusing the system libraries names when compiling boost from source? I think that way round it would really be the safer option (e.g. prefix everything with `sc_`).

> I'm not sure what you mean... do you mean link against boost or use its include directory? the comment i left about is about a call to target_include_directories afaict

Must be Sunday... oh it is... nevermind :>

brianlheim commented:
> Is it possible to set new lib names (e.g. ${sc_boost_system}) for the system libraries using add_library(), instead of reusing the system libraries names when compiling boost from source? I think that way round it would really be the safer option (e.g. prefix everything with sc_).

I agree. we could use https://cmake.org/cmake/help/v3.3/command/add_library.html#alias-libraries

> However, I like the unifying approach and this will make things more readable, but I really don't know what will happen, if implemented (as it needs build tests against several targets - which should also be done by travis imho) and specifically, if this should be done in this pull request as well.

yes we should have at least 1 travis target to do this, preferably one of the linux ones. i can help with that if you want.

yes it should be done in this PR, i don't think it'll be too much work and it's going to be about the same diff as maintaining the current tangle (with less review time overall). you make a good point that really it's hard to verify this is working the way we want unless we have reproducible CI for it. since this is seeming to be a big problem lately we should do that ASAP to save everyone time.

dvzrv commented:
Okay, will try that out locally and rebuild against the targets I have: with system boost and system yaml-cpp, with system boost and vendor yaml-cpp (vendor boost and system yaml-cpp/vendor yaml-cpp is somewhat blocked by #4183 for me currently).

dvzrv commented on CMakeLists.txt:
```diff
@@ -253,14 +253,15 @@ if (NOT WIN32)
+    set(Boost_USE_MULTITHREADED ON)
+    find_package( Boost 1.50.0 COMPONENTS  thread system filesystem program_options regex test_exec_monitor )
+    if (Boost_FOUND)
+        link_directories(${Boost_LIBRARY_DIRS})
```
Apparently it wasn't! :)

dvzrv commented:
Hmm, I think I will do the travis integration in a separate pull request.
Working on it, I changed, simplified and fixed many of the scripts below .travis and changed the way .travis.yml is defining which builds are using Qt. I think this is better reviewed separately!

brianlheim commented:
@dvzrv Typically we only dismiss reviews when they are blocking a merge.

I will try to review this this week, thanks for putting in the effort!

dvzrv commented:
Urgh, sorry, I thought there was something wrong with github in this case (I couldn't see any more open requests, but the review was showing as still open). Didn't realize you never closed it (and thought this had something to do with the requests in there). No bad intentions on my side. My bad!

dvzrv commented:
@brianlheim is there a chance we can merge this soon? I would like to continue with the travis integration for it.

dvzrv commented:
@snappizz @brianlheim are there any further things to change? It would be great to merge this.

brianlheim commented on CMakeLists.txt:
```diff
@@ -253,20 +253,43 @@ if (NOT WIN32)
+    else()
+        set(SYSTEM_BOOST OFF)
+        set(boost_include_dirs ${CMAKE_SOURCE_DIR}/external_libraries/boost)
+        message(WARNING "Not using system boost because none was found.")
```
i think this should be a FATAL_ERROR; i can't think of a situation where this is what the user actually wants to happen

dvzrv replied:
You're probably right. I thought this would be a nice "fallback", but explicitely failing is better, indeed.

dvzrv commented:
@brianlheim This worked on all the iterations I tried this with (with system boost/system yamlcpp, system boost/ vendor yamlcpp, vendor boost/ system yamlcpp, vendor boost/ vendor yamlcpp)
